core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -285,14 +296,15 @@ protected void recoverFromTranslog(EngineConfig engineConfig, Translog.TranslogG
     }
 
     private SearcherManager createSearcherManager() throws EngineException {
-        boolean success = false;
+        DirectoryReader directoryReader = null;
         SearcherManager searcherManager = null;
         try {
             try {
-                final DirectoryReader directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter, true), shardId);
+                directoryReader = ElasticsearchDirectoryReader.wrap(DirectoryReader.open(indexWriter, true), shardId);
                 searcherManager = new SearcherManager(directoryReader, searcherFactory);
-                lastCommittedSegmentInfos = readLastCommittedSegmentInfos(searcherManager, store);
-                success = true;
+                if (lastCommittedSegmentInfos == null) {
```

Comment:
This feels a bit weird. Maybe we should separate this into two methods:
1. create the reader and read the lastComittedInfos
2. create a searcherManger based on the given reader. If one needs a dedicated directoryReader per searcher manager (not sure of lucene semantics), can we maybe call  `lastCommittedSegmentInfos = readLastCommittedSegmentInfos(searcherManager, store);` once in the constructor?